### PR TITLE
Distribute first points in a BalancingLearner

### LINKED ISSUE #159: BalancingLearner puts all points in the first child-learner when asking for points with no data present
```python
def h(x, offset=0):
    a = 0.01
    return x + a**2 / (a**2 + (x - offset)**2)

learners = [adaptive.Learner1D(partial(h, offset=random.uniform(-1, 1)),
            bounds=(-1, 1)) for i in range(10)]

bal_learner = adaptive.BalancingLearner(learners)

bal_learner.ask(10)
```
```
([(0, -1),
  (0, 1),
  (0, 0.0),
  (0, -0.5),
  (0, 0.5),
  (0, -0.75),
  (0, -0.25),
  (0, 0.25),
  (0, 0.75),
  (0, -0.875)],
 [inf, inf, inf, inf, inf, inf, inf, inf, inf, inf])
```

Instead, it should balance the points over all learners. This happens because the loss improvements are `inf` for all points.

A temporary solution would be to force all learners to have the same amount of points by using:
```
bal_learner = adaptive.BalancingLearner(learners, strategy='npoints')
```

### PR DESCRIPTION
**Pull request by basnijholt**
This fixes #159.

@akhmerov could you take a look at this?

### CODE REVIEW

**Comment by akhmerov** [comment]
This fix seems to be only partial: consider for example the case when one learner is much cheaper than all the others. Then as soon as one point arrives, the balancing learner will request all points from another random learner in the queue.

If I understand correctly, an empty learner has both an infinite loss and an infinite loss improvement, however some non-empty learners also have the same loss and loss improvement. In that case the problem would arise as well.

Therefore it seems to suggest that a better fix would be to sort on `(-loss, n_points)` (or loss improvement, depending on the balancing learner setting).

**Comment by basnijholt** [comment]
@akhmerov, good suggestion! 

I've done this in 6e172ef85228e5f4469488fe07d6ae0773b4b498. This is what you meant right?

I'll also do it in `_ask_and_tell_based_on_loss`.

File: adaptive/learner/balancing_learner.py [hunk 1]
```diff
@@ -112,25 +112,33 @@ def strategy(self, strategy):
...
-            x, l = max(zip(pairs, improvements_per_learner),
-                       key=itemgetter(1))
-            points.append(x)
-            loss_improvements.append(l)
-            self.tell_pending(x)
-
-        return points, loss_improvements
+                points, loss_improvements = self._points[index]
+                npoints = npoints_per_learner[index] + learner.npoints
+                sort_tuple = (loss_improvements[0], -npoints)
```
Comment (akhmerov) [inline adaptive/learner/balancing_learner.py]: `priority`?

File: adaptive/learner/balancing_learner.py [hunk 2]
```diff
@@ -112,25 +112,33 @@ def strategy(self, strategy):
...
+                improvements_per_learner.append(sort_tuple)
+                points_per_learner.append((index, points[0]))
+
+            # Chose the optimal improvement.
+            (index, point), (loss_improvement, _) = max(
+                zip(points_per_learner, improvements_per_learner),
+                key=itemgetter(1))
+            npoints_per_learner[index] += 1
```
Comment (akhmerov) [inline adaptive/learner/balancing_learner.py]: Does this ever get decremented?

Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: No it won't.

Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: No it won't.

Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: But then what does it mean? Above I see `npoints_per_learner[index] + learner.npoints`

Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: learner.npoints doesn't change when we add a point to chosen_points, so without this all suggested points would come from the same learner.

Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: learner.npoints doesn't change when we add a point to chosen_points, so without this all suggested points would come from the same learner.

Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: But shouldn't we decrement this when a point is added to the learner? Also what is the meaning of this variable? It's not at all clear from how it's used, and seems to duplicate `learner.npoints`.

Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: but `learner.npoints` doesn't increase when adding something to `tell_pending`.

Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: So in this loop `for index, learner in enumerate(self.learners):` we ask a point from each learner. After this we can pick the `learner` and `point` with the highest `loss_improvement`, this means that in our `chosen_points` list we will have one point for the `learner` we chose. However, `learner.npoints` wouldn't have increased at this point in time! So the next time we execute this loop we need to take the `chosen_points` into account.

Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: Ah! I overlooked that this isn't an attribute.

**Comment by akhmerov** [comment]
My problem right now is that the logic is rather nontrivial, and it's hard to follow especially due to very similar variable names. Comments might also help.

File: adaptive/learner/balancing_learner.py [hunk 3]
```diff
@@ -111,38 +111,57 @@ def strategy(self, strategy):
             raise ValueError(
                 'Only strategy="loss_improvements", strategy="loss", or'
                 ' strategy="npoints" is implemented.')
+        self._points = {}  # reset the cache
 
     def _ask_and_tell_based_on_loss_improvements(self, n):
-        points = []
-        loss_improvements = []
+        chosen_points = []
```
Comment (akhmerov) [inline adaptive/learner/balancing_learner.py]:
```python
selected = []  # tuples (learner_index, point) 
```

File: adaptive/learner/balancing_learner.py [hunk 4]
```diff
@@ -111,38 +111,57 @@ def strategy(self, strategy):
...
+        chosen_loss_improvements = []
+        npoints_per_learner = defaultdict(int)
```
Comment (akhmerov) [inline adaptive/learner/balancing_learner.py]: Rename to `n_selected` or something like that

Add a comment along the lines of "`same as Counter(i[0] for i in selected)`"

File: adaptive/learner/balancing_learner.py [hunk 5]
```diff
@@ -111,38 +111,57 @@ def strategy(self, strategy):
             raise ValueError(
                 'Only strategy="loss_improvements", strategy="loss", or'
                 ' strategy="npoints" is implemented.')
+        self._points = {}  # reset the cache
 
     def _ask_and_tell_based_on_loss_improvements(self, n):
-        points = []
-        loss_improvements = []
+        chosen_points = []
+        chosen_loss_improvements = []
```
Comment (akhmerov) [inline adaptive/learner/balancing_learner.py]: Just `loss_improvements`?

File: adaptive/learner/balancing_learner.py [hunk 6]
```diff
@@ -111,38 +111,57 @@ def strategy(self, strategy):
...
-                pairs.append((index, point[0]))
-            x, l = max(zip(pairs, improvements_per_learner),
-                       key=itemgetter(1))
-            points.append(x)
-            loss_improvements.append(l)
-            self.tell_pending(x)
-
-        return points, loss_improvements
+                points, loss_improvements = self._points[index]
+                npoints = (npoints_per_learner[index]
```
Comment (akhmerov) [inline adaptive/learner/balancing_learner.py]: `total_points` or just `total`

File: adaptive/learner/balancing_learner.py [hunk 7]
```diff
@@ -111,38 +111,57 @@ def strategy(self, strategy):
...
-                improvements_per_learner.append(loss_improvement[0])
-                pairs.append((index, point[0]))
-            x, l = max(zip(pairs, improvements_per_learner),
-                       key=itemgetter(1))
-            points.append(x)
-            loss_improvements.append(l)
-            self.tell_pending(x)
-
-        return points, loss_improvements
+                points, loss_improvements = self._points[index]
```
Comment (akhmerov) [inline adaptive/learner/balancing_learner.py]: This is really confusing and seems like we've chosen a bad name for `self._points`: from that name it's not at all clear that loss improvements are stored, neither it is clear what these `self._points` actually are. All points ever requested?

**Comment by basnijholt** [comment]
I found another issue that I fixed. 

I've simplified it now and I think the code is understandable, do you agree?

File: adaptive/learner/balancing_learner.py (hunk 3 above)
Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: Overall we should make a clear distinction between `point` and `(learner_index, point)`. Conceptually these are the same (the latter is a point from a balancing learner perspective). This makes them easy to confuse, and therefore a clear name separation is very useful.

File: adaptive/learner/balancing_learner.py (hunk 7 above)
Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: It's an cache for `ask`, so maybe `_ask_cache`?

Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: Yes, that's much better.

File: adaptive/learner/balancing_learner.py (hunk 4 above)
Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: The code is no longer the same, so I resolved it.

File: adaptive/learner/balancing_learner.py (hunk 5 above)
Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: I call `loss_improvements` what is returned be `ask`, so always `loss_improvements = ask(n=1)[1]`.

File: adaptive/learner/balancing_learner.py (hunk 3 above)
Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: Maybe we could make `selected` like:

`selected = []  # tuples ((learner_index, point), loss_improvement) `

Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: Indeed, however shouldn't it be flattened as well then?

Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: Yes, like I did in the last commit?

File: adaptive/learner/balancing_learner.py [hunk 8]
```diff
@@ -113,50 +113,67 @@ def strategy(self, strategy):
                 ' strategy="npoints" is implemented.')
 
     def _ask_and_tell_based_on_loss_improvements(self, n):
-        points = []
-        loss_improvements = []
+        selected = [] # tuples ((learner_index, point), loss_improvement)
+        npoints = [l.npoints + len(l.pending_points) for l in self.learners]
```
Comment (akhmerov) [inline adaptive/learner/balancing_learner.py]: Maybe `total_points` (to indicate that it's evaluated and pending together)?

File: adaptive/learner/balancing_learner.py [hunk 9]
```diff
@@ -113,50 +113,67 @@ def strategy(self, strategy):
...
         for _ in range(n):
             improvements_per_learner = []
-            pairs = []
+            points_per_learner = []
```
Comment (akhmerov) [inline adaptive/learner/balancing_learner.py]: These aren't per learner (I'd only call something "per learner" if it's a list of length equal to the number of learners).

Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: Are these `requested`?

File: adaptive/learner/balancing_learner.py (hunk 3 above)
Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: Yes, but what about flattening this data structure as well?

Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: NB: PEP8 suggests a double space before a trailing comment.

File: adaptive/learner/balancing_learner.py (hunk 9 above)
Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: I've changed it. What do you think?

File: adaptive/learner/balancing_learner.py (hunk 8 above)
Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: Done.

File: adaptive/learner/balancing_learner.py (hunk 3 above)
Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: Fixed!

File: adaptive/learner/balancing_learner.py (hunk 6 above)
Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: Done

File: adaptive/learner/balancing_learner.py (hunk 3 above)
Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: Why not also flatten, so `(learner_index, point, loss_improvement)`?

Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: Disregard that; it's the format the runner expects.

File: adaptive/learner/balancing_learner.py [hunk 10]
```diff
@@ -113,54 +113,72 @@ def strategy(self, strategy):
...
-            self.tell_pending(x)
-
+                points, loss_improvements = self._ask_cache[index]
+                to_select.append(
+                    ((index, points[0]),
+                     (loss_improvements[0], -total_points[index]))
+                )
+
+            # Choose the optimal improvement.
+            (index, point), (loss_improvement, _) = max(
```
Comment (akhmerov) [inline adaptive/learner/balancing_learner.py]: Shouldn't we remove the point from `_ask_cache` around here?

Comment (basnijholt) [reply adaptive/learner/balancing_learner.py]: Very well spotted!

I've added `self.tell_pending((index, points[0]))`.

File: adaptive/learner/balancing_learner.py (hunk 9 above)
Comment (akhmerov) [reply adaptive/learner/balancing_learner.py]: Much better now. I think a reader can now understand what's happening without looking elsewhere.

File: adaptive/tests/test_balancing_learner.py [hunk 11]
```diff
@@ -21,3 +22,34 @@ def test_balancing_learner_loss_cache():
     bl = BalancingLearner([learner])
     assert bl.loss(real=False) == pending_loss
     assert bl.loss(real=True) == real_loss
+
+
+def test_distribute_first_points_over_learners():
+    for strategy in ['loss', 'loss_improvements', 'npoints']:
```
Comment (akhmerov) [inline adaptive/tests/test_balancing_learner.py]: `pytest.mark.parameterize`? (same below)

File: adaptive/tests/test_balancing_learner.py [hunk 12]
```diff
@@ -21,3 +22,34 @@ def test_balancing_learner_loss_cache():
...
+        learners = [Learner1D(lambda x: x, bounds=(-1, 1)) for i in range(10)]
+        learner = BalancingLearner(learners, strategy=strategy)
+        points, _ = learner.ask(100)
```
Comment (akhmerov) [inline adaptive/tests/test_balancing_learner.py]: We should also test for a failure mode when e.g. 1 or 3 points are fed into a learner.

File: adaptive/tests/test_balancing_learner.py (hunk 11 above)
Comment (basnijholt) [reply adaptive/tests/test_balancing_learner.py]: Good one. Done!

File: adaptive/tests/test_balancing_learner.py (hunk 12 above)
Comment (basnijholt) [reply adaptive/tests/test_balancing_learner.py]: Done in 6e3e9852e828ee5f5d016f26a4d1e2453c606aaf.
